java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/htmlui/Browser.java

```diff
@@ -0,0 +1,741 @@
+     *
+     * @throws Exception
+     */
+    public Browser() {
+        this(new Config());
+    }
+
+    /**
+     * Browser configured by provided config.
+     *
+     * @param config the configuration
+     */
+    public Browser(Config config) {
+        this(findCalleeClassName(), config, null);
+    }
+
+    Browser(String app, Config config, Supplier<HttpServer<?,?,?, ?>> serverProvider) {
+        this.serverProvider = serverProvider != null ? serverProvider : () -> {
+            return new SimpleServer(config.random);
+        };
+        this.app = app;
+        this.config = new Config(config);
+    }
+
+    @Override
+    public final void execute(final Runnable r) {
+        current.execute(r);
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (server != null) {
+            server.shutdownNow();
+        }
+    }
+
+    HttpServer server() {
+        return server;
+    }
+
+    static HttpServer findServer(Object obj) {
+        Command c;
+        if (obj instanceof Command) {
+            c = (Command) obj;
+        } else if (obj instanceof ProtoPresenter) {
+            c = ((ProtoPresenter) obj).lookup(Command.class);
+        } else {
+            throw new IllegalArgumentException("Cannot find server for " + obj);
+        }
+        return c.browser.server();
+    }
+
+    /** Shows URL in a browser.
+     * @param page the page to display in the browser
+     * @throws IOException if something goes wrong
+     */
+    void show(URI page) throws IOException {
+        config.getBrowser().accept(page);
+    }
+    @Override
+    public Fn defineFn(String string, String... strings) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void loadScript(Reader reader) throws Exception {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Fn defineFn(String string, String[] strings, boolean[] blns) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void flush() throws IOException {
+        throw new UnsupportedOperationException();
+    }
+
+    private static URI pageURL(String protocol, HttpServer server, final String page) {
+        int port = server.getPort();
+        try {
+            return new URI(protocol + "://localhost:" + port + page);
+        } catch (URISyntaxException ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    @Override
+    public final void displayPage(URL page, Runnable onPageLoad) {
+        try {
+            this.onPageLoad = onPageLoad;
+            this.server = serverProvider.get();
+            int from = 8080;
+            int to = 65535;
+            int port = config.getPort();
+            if (port != -1) {
+                from = to = port;
+            }
+            server.init(from, to);
+
+            String prefix = "/" + new UUID(config.random.nextLong(), config.random.nextLong()).toString();
+
```

Comment:
https://github.com/apache/netbeans-html4j/pull/26#issuecomment-981762288 questions the security when opening a port. The 24cb4da fix is supposed to make sure the port is only available locally. Moreover e91bdcb uses random UUID to prefix the server pages and in case of misconnect, it closes the server. 

That shall mitigate the basic problem. We can treat the rest as a bug.

Reply:
a65a7de synchronizes the `SimpleServer` class with https://github.com/apache/netbeans-html4j/pull/26/commits/9bc1429686ea37037bba33d9d3a8123ae9ff1148